viewing_party/party.py

```diff
@@ -1,23 +1,159 @@
 # -----------------------------------------
+def get_unique_watched(user_data):
+# returns a list of dict movies the user has watched, but none of their friends have watched
+    unique_movies = []
+    all_friends_movies = [] 
+
+    for friend in user_data["friends"]:
+        for movie in friend["watched"]:
+            if movie not in all_friends_movies:
+                all_friends_movies.append(movie)
+
+    for movie in user_data["watched"]:
+        if movie not in all_friends_movies:
+            unique_movies.append(movie)
+    return unique_movies
+
+
+def get_friends_unique_watched(user_data):
+    # Returns list of dicts movies at least one of the user's friends have watched, but the user has not watched
+    user_movies = []
+    unique_friend_movies = []
+    for movie in user_data["watched"]:
+        if movie not in user_movies:
+            user_movies.append(movie)
+
+    for friend in user_data["friends"]:
+        for movie in friend["watched"]:
+            if movie not in user_movies and movie not in unique_friend_movies:
+                unique_friend_movies.append(movie)
+    return unique_friend_movies
+
+
 
-        
 # -----------------------------------------
 # ------------- WAVE 4 --------------------
 # -----------------------------------------
+def get_available_recs(user_data):
+    # Returns a list of recommended movies that the user has not watched, at least one of the user's friends has watched
+    # and the host of the movies has to be in the users subscription 
+    list_rec_movies = []
+    sub_list = []
+    unique_watched = get_friends_unique_watched(user_data)
+    
+    for sub in user_data["subscriptions"]:
+        sub_list.append(sub)
+            
+    for movie in unique_watched:
+        if movie["host"] in sub_list:
+            list_rec_movies.append(movie)
+                
+    return list_rec_movies
 
 # -----------------------------------------
 # ------------- WAVE 5 --------------------
 # -----------------------------------------
+def get_new_rec_by_genre(user_data):
+    # Returns a list of recommended movies that the user has not watched, at least one of the user's friends has watched
+    # and the genre of the movie is the same as the most frequent genre
+    rec_movies_list = []
+    popular_genre = get_most_watched_genre(user_data)
+    unique_watched = get_friends_unique_watched(user_data)
+
+    if not user_data["watched"]:
+        return rec_movies_list
+    for movie in unique_watched:
+            if movie["genre"] in popular_genre:
+                rec_movies_list.append(movie)
+    return rec_movies_list
+    
+def get_rec_from_favorites(user_data):
+    # Returns a list of recommended movies if it's in the user's favorite 
+    # and non of the user's friends have watched it
+    fav_rec_movies = []
+    friends_watched = []
+    
+    for movie in user_data["friends"]:
+        for item in movie["watched"]:
+            friends_watched.append(item)

```

Comment:
This nested for loop is similar to code above, so just a reminder that it could have been a helper function!